addons/l10n_id_efaktur/models/efaktur.py

```diff
@@ -12,46 +12,52 @@ class Efaktur(models.Model):
     _description = "Available E-faktur range"
 
     company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
-    max = fields.Char(compute='_compute_default', store=True, readonly=False)
-    min = fields.Char(compute='_compute_default', store=True, readonly=False)
-    available = fields.Integer(compute='_compute_available', store=True)
+    max = fields.Char(required=True)
+    min = fields.Char(required=True)
+    available = fields.Integer(readonly=True)
+    next_num = fields.Integer(compute="_compute_next_num")
 
-    @api.model
-    def pop_number(self, company_id):
-        range = self.search([('company_id', '=', company_id)], order="min ASC", limit=1)
-        if not range:
-            return None
-
-        popped = int(range.min)
-        if int(range.min) >= int(range.max):
-            range.unlink()
-        else:
-            range.min = '%013d' % (popped + 1)
-        return popped
-
-    @api.model
-    def push_number(self, company_id, number):
-        return self.push_numbers(company_id, number, number)
+    @api.depends('max', 'available')
+    def _compute_next_num(self):
+        for record in self:
+            record.next_num = int(record.max) - record.available + 1
 
-    @api.model
-    def push_numbers(self, company_id, min, max):
-        range_sup = self.search([('min', '=', '%013d' % (int(max) + 1))])
-        if range_sup:
-            range_sup.min = '%013d' % int(min)
-            max = range_sup.max
+    def pop_number(self):
 
-        range_low = self.search([('max', '=', '%013d' % (int(max) - 1))])
-        if range_low:
-            range_sup.unlink()
-            range_low.max = '%013d' % int(max)
+        popped = self.next_num
+        self.available -= 1
 
-        if not range_sup and not range_low:
-            self.create({
-                'company_id': company_id,
-                'max': '%013d' % int(max),
-                'min': '%013d' % int(min),
-            })
+        return popped
 
+    @api.model
+    def push_number(self, company_id, number):
+        efaktur_range = self.search([('company_id', '=', company_id), ('min', '<=', number), ('max', '>=', number)])
+
+        # if released number is the last released number from a range, simply extend the availability
+        if efaktur_range.next_num == int(number) + 1:
+            efaktur_range.available += 1
+            return
+
+        # if the released number is not the last used number from any range, find the range it belongs and split it
+        # i.e range: 1 - 10, available=5, next_number=6, release 3
+        # split 1-10 to 1-3 available=1 AND 4-10 available=5
+        if efaktur_range:
```

Comment:
What would happen if a user post migration release a number that does not belong to a range?
Maybe we should support this use case, at least temporarily :thinking:

Reply:
I guess in this case we can create a range specific to that number (i.e. release 5 => create range with min=5, max=5, available=1)?